src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  const addProperties = (extraData) => {
+    stateCopy = Object.assign(stateCopy, extraData);
+    stateHistory.push({ ...stateCopy });
+  };
+
+  const removeProperties = (keysToRemove) => {
+    for (const remove in keysToRemove) {
```

Comment:
The 'for...in' loop is not the correct choice for iterating over an array. You should use a simple 'for' loop, 'for...of' loop, or array methods like 'forEach' for array iteration.